### PR TITLE
add extension point for custom StringMatcher, and lua implementation

### PR DESCRIPTION
**Pull request by ggreenway**
Commit Message: Add an extension point for custom StringMatcher implementations, with a Lua extension
Additional Description:
Risk Level: Low
Testing: New tests
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by ggreenway** [comment]
Note: I pulled one commit into a separate PR (#32587) to make it easier to review.

File: api/envoy/type/matcher/v3/string.proto
Comment (htuch) [inline api/envoy/type/matcher/v3/string.proto:68]: We probably should in general be using this anyway, Should we have some CI format check to ensure we don't use stale xDS types  @envoyproxy/api-shepherds?

File: api/envoy/type/matcher/v3/string.proto
Comment (htuch) [inline api/envoy/type/matcher/v3/string.proto:69]: @ggreenway what's the main difference between this and using the `safe_regex` field with a custom matcher there?

Comment (htuch) [reply api/envoy/type/matcher/v3/string.proto:69]: Answering my own extension, we only have `oneof`, not pluggable extensions for `RegexMatcher`.

File: api/envoy/extensions/matching/string/lua/v3/lua.proto
Comment (htuch) [inline api/envoy/extensions/matching/string/lua/v3/lua.proto:20]: Should this be a `DataSource`? Comparing with the HTTP Lua equivalent.

File: api/envoy/extensions/matching/string/lua/v3/lua.proto
Comment (htuch) [inline api/envoy/extensions/matching/string/lua/v3/lua.proto:3]: FWIW the `envoy.extensions.matching` namespace is today used for the generic matcher input matchers. These are a very specific kind of match use case. I might be tempted to put this in `envoy.extensions.string_matcher`, but not that opinionated if we want to expand the matching hierarchy use cases. CC @envoyproxy/api-shepherds

File: api/envoy/type/matcher/v3/string.proto
Comment (wbpcode) [inline api/envoy/type/matcher/v3/string.proto:69]: prefer the name `custom` to mean this is a custom matching.

File: api/envoy/type/matcher/v3/string.proto
Comment (wbpcode) [reply api/envoy/type/matcher/v3/string.proto:68]: Hmmm, there also a `xds.type.matcher.v3.StringMatcher` in xds repo, should we also update it?

@htuch Now the stale xDS types are used in almost everywhere. And there are lots of new APIs  (which are not synced to xds repo) in the `envoy/config/core` and `envoy/type/matcher`, so I think we should re-think does the xds repo still has actual meaning in practice.

cc @envoyproxy/api-shepherds

Comment (markdroth) [reply api/envoy/type/matcher/v3/string.proto:68]: I believe that our intention was to migrate to the types in the `xds` namespace over time.  I think we probably do need to enforce this somehow, or it will never happen.  We should stop allowing new instances of the old types in the `envoy` namespace.  We should also consider saying that new fields should be added only to the `xds` types, as a way of promoting migration: if a given call site needs a new field, that's an opportunity to add a field for the `xds` matcher type as an alternative to the existing `envoy` type.

File: api/envoy/type/matcher/v3/string.proto
Comment (ggreenway) [reply api/envoy/type/matcher/v3/string.proto:69]: Because StringMatcher is used in many places where the new matchers are not supported, and adding this here gets support in all those places.

I'm not very familiar with the new matcher API, but I looked at it briefly and it seems to always have actions associated with a match; I'm not exactly sure how to apply that in some cases where StringMatcher is used where it just needs a boolean response.

One specific case I want to use this extension matcher is in SAN validation in the tls transport socket.

File: api/envoy/type/matcher/v3/string.proto
Comment (ggreenway) [reply api/envoy/type/matcher/v3/string.proto:68]: @wbpcode yes I think it makes sense to also add it to the xds StringMatcher type; that can be done pretty easily later.

Given how pervasive both the StringMatcher and TypedExtensionConfig are in the envoy namespace in the envoy config, I don't think we're ever going to be able to migrate existing usage to the xds namespace. It would break literally every envoy config, so I think we'll have to support using both. Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Comment (wbpcode) [reply api/envoy/type/matcher/v3/string.proto:68]: I think it must be a reason that force us to use do this migration? But I didn't get it.

As a developer of Envoy, I think it's much easier to use/update the APIs under Envoy. Or we always to update another repo (I mean the cncf/xds repo).
As a reviewer, I think it would be more friendly to make the API and the implementation in same PR.
As a developer of control plane (a private fork of istio), I always use the `envoyproxy/go_control_plane` and never care the `cncf/xds`.
As a user of Envoy, I hope all APIs could be placed in the same position.
And I believe even as a developer of another xds client (like grpc), it's impossible to just use the `cncf/xds` only because it's simple and lack of good maintenance (compared to `envoy/api`).

In conclusion, I don't get a point why the migration is necessary. Anyone know the actual background and reason?

cc @envoyproxy/api-shepherds @envoyproxy/senior-maintainers

Comment (wbpcode) [reply api/envoy/type/matcher/v3/string.proto:68]: > It would break literally every envoy config, so I think we'll have to support using both. Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Yeah. I agree with that migration is impossible.

And I also inclined to use `envoy.config.core.v3.TypedExtensionConfig` here (`envoy.type.matcher.v3.StringMathcer`) directly and use `xds.core.v3.TypedExtensionConfig` in `xds.type.matcher.v3.StringMatcher`. If we actually want to use the xds one to replace the envoy one, we should use the whole `xds.type.matcher.v3.StringMacher` to replace `envoy.type.matcher.v3.StringMatcher`.

It's weird for me to add a new field that use the xds API here because I think the `StringMatcher` is a very basic type.

The circular dependency is a problem, but maybe we could resolve in other way.

Comment (htuch) [reply api/envoy/type/matcher/v3/string.proto]: >  Forcing new usage to the xds namespace seems fine, but migration seems impossible.

Agree, we will followup at some point with tooling to enforce this as presubmit.

@wbpcode we will have a proposal coming up to move more things to the xDS repository. This would enable multi data plane consumption of xDS in a credible way with proper maintainership and governance. We will bring this to a GitHub issue and Envoy community meeting in the near future. I 100% appreciate the concerns you raise and I'm happy to do an offline 1:1 on this topic - I've started on a one pager to share for Envoy maintainers.

In the meanwhile, we should probably maintain our existing posture, which is to prefer these the cncf/xds base types where possible, so at least in this PR, we're all good. 

Thanks for the discussion and input.

File: api/envoy/type/matcher/v3/string.proto
Comment (htuch) [reply api/envoy/type/matcher/v3/string.proto]: Yeah, I think I'm fine conceptually here with this being an extension type for the reasons Greg gives.

**Comment by htuch** [comment]
> Hmmm, there also a xds.type.matcher.v3.StringMatcher in xds repo, should we also update it?

@wbpcode this is a really good point. When we mirror types into the xds package namespace, we then have the challenge of forcing canonical evolution of those types to take place there and need a systematic way to bring them back into the legacy types for cases where features like this matter.

This is actually pretty messy today as this example shows. I think for this reason, I'd discourage this pattern in the future - regardless of the repository where a proto lives, we should stick to ODR and not replicate across package namespaces, at least if that type has some material amount of usage in the API already.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: source/server/server.cc
Comment (htuch) [inline source/server/server.cc:111]: Why was this needed?

File: source/common/common/matchers.h
Comment (htuch) [inline source/common/common/matchers.h:142]: As per @wbpcode should we update this as well to keep consistency?

**Review by htuch — COMMENTED** [review]
LGTM modulo comments.

File: source/common/common/matchers.h
Comment (ggreenway) [reply source/common/common/matchers.h:142]: Yes, I think we can update that as well (in a separate commit), and then remove some of the template/overloading code that makes it work with the types not being identical.

File: source/server/server.cc
Comment (ggreenway) [reply source/server/server.cc:111]: This is needed to make the multi-envoy integration test pass. Without this, the 2nd server ASSERTs because it's trying to set a non-null singleton.

I think this means that extensions that need this singleton will not work in certain envoy-mobile setups where there are multiple engines running.

As best I can tell, this is the same situation we're in for a [regex engine](https://github.com/envoyproxy/envoy/blob/28ed0f4f3fba491bbf0abc2bcfa69430b501f2ad/source/server/regex_engine.cc#L23C3-L24C58), which is also managed by a singleton (for identical reasons: it's REALLY ugly to pass all these bits in to every place where a StringMatcherImpl is constructed; I tried, and it touched too much code).

File: source/common/common/matchers.h
Comment (htuch) [reply source/common/common/matchers.h:142]: SG.

File: source/server/server.cc
Comment (htuch) [reply source/server/server.cc:111]: OK, maybe leave a comment on this? It's unclear to me which things are chosen for initialization here and why other than trial-and-error.

Comment (ggreenway) [reply source/server/server.cc:118]: Agreed; adding a comment

Comment (htuch) [reply source/server/server.cc:118]: Thanks, I can see the issue from this. I think it might be OK to selectively allow this violation of singleton safety, but I'd like to just quickly check with @mattklein123 (who did a lot of the singleton work) and @yanavlasov (who has coverage on how we create Google internal environments for Envoy binaries) to verify. Good to merge if we can agree on this.

Comment (KBaichoo) [reply source/server/server.cc:118]: Is this still pending @ggreenway ?

Comment (ggreenway) [reply source/server/server.cc:118]: Yep, still waiting on input/signoff.

Comment (htuch) [reply source/server/server.cc:118]: Friendly ping @mattklein123 @yanavlasov.

Comment (yanavlasov) [reply source/server/server.cc:118]: It would be hard to tell what impact it will have internally. I think the code is contained enough that I'm not too concerned about patching it internally.

My main concern here is that a bad precedent with regex engine is getting reinforced here. If a factory does not provide the context that an extension needs, people will continue plumbing missing objects through a singleton. 

I think it may be ok to accept this as a temporary solution. However I think the matcher factory needs to be fixed to have at least the base (or server) context passed to it. I suspect it is a major pain.

This should also be a requirement that all new factories have context available.

Comment (mattklein123) [reply source/server/server.cc:118]: I agree with @yanavlasov. This is OK for now but it's not great and we should try to track cleaning it up as tech debt. In a perfect world it would be nice for use of this to fail in the multi-envoy case vs. just not work silently but I'm not sure if that is easily possible.

Comment (ggreenway) [reply source/server/server.cc:118]: I agree it's not ideal. Before posting the PR, I started trying to pass the context through everywhere and the change was HUGE and I hadn't yet solved the most awkward case, which is that the TLS cert validator gets constructed from an SDS secret-update callback, and we'd need the context there.

Filed #32792 to track.

**Review by htuch — APPROVED** [review]

**Pull request merged**
